File: server/Socks/Packet.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Square;

namespace CircleScape {
    class Packet {
        public enum kId {
            KeyExchange = 0,
            LoginAttempt,
            RegistrationAttempt
        }
        
        private static Packet ErrorPacket {
            get => new Packet { IsLegal = false };
        }

        public static Packet FromBytes(byte[] raw) {
            if(raw.Length < 3)
                return ErrorPacket;

            Packet packet = new Packet();
            if(!Enum.IsDefined(typeof(kId), (int)raw[0]))
                return ErrorPacket;
            packet.Id = (kId)raw[0];
            var regionCount = raw[1];
            var regionLengths = new List<uint>();
            var headerPtr = 2;
            for(var i = 0; i < regionCount; ++i) {
                regionLengths.Add(0);
                var first = raw[headerPtr];
                if(first < 254) {
                    regionLengths[i] = first;
                    ++headerPtr;
                } else if(first == 254) {
                    if(headerPtr + 3 < raw.Length)
                        return ErrorPacket;
                    regionLengths[i] = raw.Subset(headerPtr + 1, 2).UnpackUInt16();
                    headerPtr += 3;
                } else {
                    if(headerPtr + 5 < raw.Length)
                        return ErrorPacket;
                    regionLengths[i] = raw.Subset(headerPtr + 1, 4).UnpackUInt32();
                    headerPtr += 5;
                }

                if(headerPtr > raw.Length)
                    return ErrorPacket;
            }

            if(headerPtr + regionLengths.Sum(x => x) > raw.Length)
                return ErrorPacket;
             
            long bodyPtr = headerPtr;
            foreach(var regionLength in regionLengths) {
                // FLAG this could fail if one region exceeds 2^31-1 in size, check later
                packet.Regions.Add(raw.Subset((int)bodyPtr, (int)regionLength));
                bodyPtr += regionLength;
            }

            return packet;
        }
        
        private List<byte[]> Regions = new List<byte[]>();
        public kId Id { get; private set; } = kId.KeyExchange;
        public bool IsLegal { get; private set; } = true;
        public int RegionCount {
            get => Regions.Count;
        }

        private Packet() { }

        public Packet(kId id, params object[] regions) {
            Id = id;

            foreach(var region in regions)
                AddRegion(region);
        }

        public Region this[int i] {
            get => new Region(Regions[i]);
        }

        public Packet AddRegion(object region) {
            if(region.GetType() == typeof(byte[]))
                Regions.Add((byte[])region);
            else if(region.GetType() == typeof(string))
                Regions.Add(((string)region).GetBytes());

            return this;
        }

        public byte[] GetBytes() {
            if(!IsLegal)
                return null;

            var header = new List<byte> {
                (byte)Id,
                (byte)RegionCount
            };
            IEnumerable<byte> body = new byte[0];
            foreach(var region in Regions) {
                if(region.Length < 0xFE)
                    header.Add((byte)region.Length);
                else if(region.Length <= 0xFFFF) {
                    header.Add(0xFE);
                    header.AddRange(((UInt16)region.Length).Pack());
                } else {
                    header.Add(0xFF);
                    header.AddRange(region.Length.Pack());
                }

                body = body.Concat(region);
            }

            return header.Concat(body).ToArray();
        }

        public class Region {
            public byte[] Data { get; private set; }

            public Region(byte[] data) {
                Data = data;
            }

            public static implicit operator byte[] (Region region) => region.Data;
            public string Bytes {
                get => this;
            }

            public static implicit operator string(Region region) {
                try {
                    return Encoding.UTF8.GetString(region.Data);
                } catch {
                    return Encoding.ASCII.GetString(region.Data);
                }
            }
            public string Str {
                get => this;
            }
        }
    }
}

File: client/src/Packet.ts
const enum kPacketId {
    KeyExchange = 0,
    LoginAttempt,
    RegistrationAttempt
}

class Packet {
    private _id: kPacketId;
    public get id(): kPacketId {
        return this._id;
    }

    private _regions: Uint8Array[] = [];
    public get regions(): Uint8Array[] {
        return this._regions;
    }
    public get regionCount(): number {
        return this._regions.length;
    }
    public getRegion(region: number): Uint8Array {
        return this._regions[region];
    }
    public getRegionString(region: number): string {
        return this.getRegion(region).toString();
    }
    public addRegion(region: object): Packet {
        if(typeof region == "string")
            this._regions.push((<string>region).toByteArray());
        else if(region instanceof Uint8Array)
            this._regions.push(region);

        this[this.regionCount-1] = this._regions[this.regionCount-1];
        return this;
    }

    private constructor() {}

    public static create(id: kPacketId, regions: any[]): Packet {
        var packet = new Packet;
        packet._id = id;
        regions.forEach(region => {
            packet.addRegion(region);
        });

        return packet;
    }

    public static fromBytes(bytes: Uint8Array): Packet {
        var packet = new Packet;
        packet._id = bytes[0];
        var regionCount = bytes[1];
        var regionLengths = [];
        var ptr = 2;
        for(var i = 0; i < regionCount; ++i) {
            if(bytes[ptr] < 0xFE)
                regionLengths.push(bytes[ptr]);
            else if(bytes[ptr] == 0xFE) {
                regionLengths.push(bytes.unpackUint16(ptr + 1));
                ptr += 2;
            } else {
                regionLengths.push(bytes.unpackUint32(ptr + 1));
                ptr += 4;
            }

            ++ptr;
        }

        for(var i = 0; i < regionCount; ++i) {
            packet.addRegion(bytes.subarray(ptr, ptr + regionLengths[i]));
            ptr += regionLengths[i];
        }

        return packet;
    }

    public getBytes(): Uint8Array {
        var headerSize = 2, bodySize = 0;
        this._regions.forEach(region => {
            bodySize += region.byteLength;

            ++headerSize;
            if(region.byteLength >= 0xFE && region.byteLength <= 0xFFFF)
                headerSize += 2;
            else if(region.byteLength > 0xFFFF)
                headerSize += 4;
        });

        var buffer = new Uint8Array(headerSize + bodySize);
        var headerPtr = 2, bodyPtr = headerSize;
        buffer[0] = this._id % 256;
        buffer[1] = this._regions.length;
        this._regions.forEach(region => {
            var regionLength = region.byteLength;
            if(regionLength < 0xFE)
                buffer[headerPtr] = regionLength;
            else if(regionLength >= 0xFE && regionLength <= 0xFFFF) {
                buffer[headerPtr] = 0xFE;
                buffer.set(regionLength.packUint16(), headerPtr + 1);
                headerPtr += 2;
            } else {
                buffer[headerPtr] = 0xFF;
                buffer.set(regionLength.packUint32(), headerPtr + 1);
                headerPtr += 4;
            }
            ++headerPtr;

            buffer.set(region, bodyPtr);
            bodyPtr += regionLength;
        });
        
        return buffer;
    }
}